# Import Python Modules

This project can be used on google Colab or an offline IDE that supports Juypter.
Check if using google Colab or vs code interpreter. If colab, py files must be imported to the colab before using import ...

In [2]:
# Import Modules

# Get path of current interpreter
import sys
interpPath = sys.executable
print("Current interpreter = " + interpPath)

# Check Path is Colab interpreter
if(interpPath == "/usr/bin/python3"):
  # Mount google drive
  from google.colab import drive
  drive.mount('/content/gdrive')

  # Import .py files from
  !cp /content/gdrive/MyDrive/University/Year\ 2/AIGP/Programs/AI-Game-Programming/Colab\ Notebooks/ActivationFunctions.py .
  !cp /content/gdrive/MyDrive/University/Year\ 2/AIGP/Programs/AI-Game-Programming/Colab\ Notebooks/ArtifcialNeuralNetwork.py .

# First Party Imports
import ActivationFunctions as act
import ArtifcialNeuralNetwork as ann

# Third Party Imports
import numpy as np
import matplotlib.pyplot as plt

Current interpreter = c:\ProgramData\anaconda3\python.exe


c:\Users\Denis\My Drive\University\Year 2\AIGP\Programs\AI-Game-Programming\Colab Notebooks\ArtifcialNeuralNetwork.py:22: SyntaxWarning: assertion is always true, perhaps remove parentheses?
  assert(INPUT_COUNT_MAX != len(X2) or INPUT_COUNT_MAX != len(YD), "ERROR- Data set pairs do not match")


# Research

## Feed forward Artificial Neural Network



### Weights
Adjust weights of neurons to provide desired outputs for all cases.

This is called supervised learning. We train neuronsusing examples. The learning algorithm maps an input to an output based on example input-output pairs.

In my code, input-output pairs are defined by X and desired outputs Y

![image.png](attachment:image.png)
![image-2.png](attachment:image-2.png)
![image-3.png](attachment:image-3.png)


### Activation Functions

![image.png](attachment:image.png)

### Inputs and Outputs

# AI Game Programming Labs

## Lab Week 4

### Code Snippets

In [8]:
# Activation Functions

# Sign Function
def Sign(X, THRESHOLD):
    if X >= THRESHOLD: return 1
    else: return -1

# Step function
def Step(X, THRESHOLD):
    if X >= THRESHOLD: return 1
    else: return 0

# Sigmoid function
def Sigmoid(X, THRESHOLD):
   return (1 / (1 + np.exp(-(X- THRESHOLD))))

def Linear(X, THRESHOLD):
    return X - THRESHOLD

In [9]:
# Weight Training Functions
# Input Function
def Input(X1,W1, X2,W2):
    return ((X1*W1) + (X2*W2))

# Error Function
def Error(YD, Y):
    return YD - Y

# Input Function
def Learn(W, LEARN_RATE, X, E):
    return (W + (LEARN_RATE * X * E))

In [10]:
# Supervised Learning Algorithm
# Threshold for activation functions
THRESHOLD = 0.2

# Learning rate for weight and error calculation
LEARN_RATE = 0.1

# Inputs and Desired Outputs
X1 = [0.0, 0.0, 1.0, 1.0]
X2 = [0.0, 1.0, 0.0, 1.0]
YD = [0.0, 0.0, 0.0, 1.0]

INPUT_COUNT_MAX = len(X1)
# NUM_INPUTS = 2

# Epoch Passes (Starts at epoch 1)
epochCount = 1
EPOCH_COUNT_MAX = 1000

# Initial input weights
w = [0.3, -0.2]
NUM_WEIGHTS = len(w)

# Feed Forward Propogation Algorithm
while(epochCount < EPOCH_COUNT_MAX):
    for i in range(0,INPUT_COUNT_MAX,1):
        # Produce input from random weights
        X = ann.Input(X1[i], w[0], X2[i], w[1])

        # Produce output from input using activation funcntion
        Y = act.Step(X, THRESHOLD)

        # Calculate error ( error = yD - Y)
        ERROR = ann.Error(YD[i], Y)

        # Adjust weights based on error
        w[0] = ann.Learn(w[0], LEARN_RATE, X1[i], ERROR)
        w[1] = ann.Learn(w[1], LEARN_RATE, X2[i], ERROR)
    epochCount += 1

# Print Results
print("Epoch Units = " + str(epochCount))
print("Learning rate = " + str(LEARN_RATE))

for i in range(0, NUM_WEIGHTS,1): print("w" + str(i) + " = " + str(w[i]))

Epoch Units = 1000
Learning rate = 0.1
w0 = 0.19999999999999998
w1 = 0.1


In [6]:
# Supervised Learning Algorithm Function
# Learning variables
THRESHOLD = 0.2
LEARN_RATE = 0.1

# Inputs and Desired Outputs
X1 = [0.0, 0.0, 1.0, 1.0]
X2 = [0.0, 1.0, 0.0, 1.0]
YD = [0.0, 0.0, 0.0, 1.0]

ann.TrainSingleIOPairs(X1, X2, YD, THRESHOLD, LEARN_RATE) # Supports additional arguments weights[2], epochs


Epoch Units = 50
Learning rate = 0.1
w0 = 0.19999999999999998
w1 = 0.1


### Task 2.1: Range of the number of epoch

Find the minimum number of epochs. Is 5 (in the example code) the minimum number of epoch
for this training?

Epoch Results

Epoch Units = 2
w0 = 0.3
w1 = -0.1

Epoch Units = 3
w0 = 0.3
w1 = 0.0

Epoch Units = 4
w0 = 0.3
w1 = 0.1

Epoch Units = 5
w0 = 0.19999999999999998
w1 = 0.1

Epoch Units = 1000
w0 = 0.19999999999999998
w1 = 0.1

Epoch Evaluation

Once Reaching a suitagbly accurate output, increasing the number of epochs to be run becomes negligble. In this case there was no difference between 5 epochs and 1000 epochs

### Task 2.2: Range of learning rate α

From the previous task, the learning rate α was set to 0.1. Systematically conduct an experiment
to find out the ‘range’ of α that works for this case
![image.png](attachment:image.png)

Learning rate Results (0.05-0.15) Yielded slightly better results

Learning rate = 0.05
Epoch Units = 1000
w0 = 0.15000000000000002
w1 = 0.04999999999999999

Learning rate = 0.06
Epoch Units = 1000
w0 = 0.18
w1 = 0.03999999999999998

Learning rate = 0.07
Epoch Units = 1000
w0 = 0.15999999999999998
w1 = 0.08000000000000002

Learning rate = 0.08
Epoch Units = 1000
w0 = 0.13999999999999996
w1 = 0.12

Learning rate = 0.09
Epoch Units = 1000
w0 = 0.12
w1 = 0.15999999999999998

Learning rate = 0.10
Epoch Units = 1000
w0 = 0.19999999999999998
w1 = 0.1

Learning rate = 0.11
Epoch Units = 1000
w0 = 0.19
w1 = 0.01999999999999999

Learning rate = 0.13
Epoch Units = 1000
w0 = 0.16999999999999998
w1 = 0.06

Learning rate = 0.14
Epoch Units = 1000
w0 = 0.15999999999999998
w1 = 0.08000000000000002

Learning rate = 0.15
Epoch Units = 1000
w0 = 0.15
w1 = 0.09999999999999998

Learning Rate Evaluation

My results showed that even a +0.1 change from the correct learning rate produced higly inaccurate weights. To reduce the inaccuracy, I tried to increase the number of epochs, hoping that this would give the algorithm a better chance of finding the correct weights. Looking back this made no difference. As shown in my epoch testing, a correct output was achieved by 4 epochs, increasing the maximum number of epochs was completely negligble past this factor.

I would suggest using a maximum range of +- 0.005 from the correct learning rate, to results close to the desired output.

### Task 2.3: Activation functions and thresholds

So far, we have used only the Step function with threshold θ = 0.2. What about other activation functions and thresholds?

 a) Step function
Systematically conduct an experiment to find out the range of theta θ that works or does not
work, in this case. Note that you need to also consider learning rate α.


b) Sign function
Systematically conduct an experiment to find out the range of theta θ that works or does not
work, in this case. Note that you need to also consider learning rate α.


c) Sigmoid function
Systematically conduct an experiment to find out the range of theta θ that works or does not
work, in this case. Note that you need to also consider learning rate α.



d) Linear function
Systematically conduct an experiment to find out the range of theta θ that works, or does not
work, in this case. Note that you need to also consider learning rate α.